006-improving-configurability-of-kafka-listeners.md

```diff
@@ -0,0 +1,240 @@
+            - broker: 2
+                advertisedHost: my-cluster-brokers-kafka-2.myns.cluster.local
+      - name: encrypted2
+        port: 9093
+        type: service
+        tls: true
+        authentication:
+          type: acram-sha-512
+      - name: routes
+        port: 9094
+        type: route
+        tls: true
+        authentication:
+          type: tls
+      - name: nodeports
+        port: 9095
+        type: nodeport
+        tls: true
+        authentication:
+          type: tls
+```
+
+The listeners will be configurable as array.
+Each listener will have its own unique name and will specify the port and type as required values.
+The port can be set to anything apart from `9091` (internal replication listener) and `9404` (Prometheus).
+A new type service will be introduced for the internal listeners designed for apps running inside the same Kubernetes cluster.
+Additionally, all the types from the existing external listeners will be supported as well.
+
+together with this change, I suggest to change the default value of the listener `tls` flag from `true` to `false`.
+The current situation, when TLS is enabled by default without using the `tls` field and needed to explicitly use the `tls: false` to disable TLS seems unintuitive and confusing.
+
+### Backwards compatibility
+
+The old format can be easily converted into the new format without any information loss.
+The example YAML below corresponds to the first example from the _Current situation_ section:
+
+```yaml
+    listeners:
+      - name: plain
+        port: 9092
+        type: service
+        tls: false
+        authentication:
+          type: scram-sha-512
+        networkPolicyPeers:
+          - podSelector:
+              matchLabels:
+                app: kafka-plaintext-consumer
+          - podSelector:
+              matchLabels:
+                app: kafka-plaintext-producer
+      - name: tls
+        port: 9093
+        type: service
+        tls: true
+        authentication:
+          type: tls
+        networkPolicyPeers:
+          - podSelector:
+              matchLabels:
+                app: kafka-consumer
+          - podSelector:
+              matchLabels:
+                app: kafka-producer
+      - name: external
+        port: 9094
+        type: route
+        tls: true
+        authentication:
+          type: tls
+```
+
+The second example could be converted like this:
+
+```yaml
+    listeners:
+      - name: plain
+        port: 9092
+        type: service
+      - name: tls
+        port: 9093
+        type: service
+```
+
+Being easily able to convert the old and new structures makes it easy to have single code in our operator while being able to easily work with the old format as well.
+
+### Code changes
+
+Since both the new and the old structure will use the same `listeners` property, the CRD generator will need to be enhanced to support the OpenAPI `oneOf` for different types.
```

Comment:
I'm sorry to have to tell you that Kubernetes does not allow CRD schema objects (subschemas if you like) to be one of a set of JSON types, see https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#schemaObject. 

To do this for v1beta1 resources we'd have to have a different property than `listeners` and use a `oneOf` constraint to force only one of them being set. This option is relatively easily done, but means we need a different name than `listeners`, and I can't think of a good one.

An alternative would be to have a `v1beta2` CRD in which `listeners` was a list as described here. That comes with a bunch of other considerations like having a upgrade webhook to convert `v1beta1` to `v1beta2` and would also require support in the `CrdGenerator` for generating different schemas. It's also predicated on dropping support for Kubernetes v1.11. This option is a *lot* more work, but results in a nicer API.

Reply:
I'm not sure I understand what do you suggest and how does the link relates. The example I listed there:

```yaml
listeners:
  oneOf:
    - type: object
        properties:
          # ...
    - type: array
        items:
          # ...
```

seemed to work fine for me.

Reply:
Schema validation is (supposedly) based on openapi v3, see https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation. Are you saying you've tried this with Kubernetes and the CRD was accepted? And it was correctly able to validate CRs using that schema?

Reply:
Oh, right I see you're not using an array-typed `type`. Maybe it would work then.

Reply:
I based it on this: https://swagger.io/docs/specification/data-models/data-types/ (section mixed types). That should refer to OAPIv3. I have a protype for the `api` and `crd-generator` changes for this. I can share it with you tomorrow. It seemed to work fine within our v1beta1 CRDs on Kube 1.18.

Reply:
OK, well if it works, that would be great, because this is something I'd really like to see.

Reply:
I used the `oneOf` for the bridge as well and the validation works (of course it's not related to Kubernetes). @tombentley out of curiosity, why it shouldn't have worked with K8S?

Reply:
I just forgot that `oneOf` (and `@OneOf`, which I implemented) didn't violate the "`type` must be scalar" rule.

Reply:
@tombentley In case you are interested ... these are the changes I have done as as part of the prototyping this (the code is of course not PR worthy, but it should give the idea): https://github.com/strimzi/strimzi-kafka-operator/compare/master...scholzj:make-it-possible-to-alternate-multiple-types-in-api-schema?expand=1